// @alfalab/core-components/date-input/index.js
Object.defineProperty(exports, '__esModule', { value: true });

var Component = require('./Component.js');
require('react');
require('../masked-input');
require('date-fns');
var utils_format = require('./utils/format.js');
var utils_dateCorrectionPipe = require('./utils/date-correction-pipe.js');
var utils_nativeSupports = require('./utils/native-supports.js');



exports.DateInput = Component.DateInput;
exports.DATE_FORMAT = utils_format.DATE_FORMAT;
exports.DATE_MASK = utils_format.DATE_MASK;
exports.NATIVE_DATE_FORMAT = utils_format.NATIVE_DATE_FORMAT;
exports.formatDate = utils_format.formatDate;
exports.isCompleteDateInput = utils_format.isCompleteDateInput;
exports.parseDateString = utils_format.parseDateString;
exports.createAutoCorrectedDatePipe = utils_dateCorrectionPipe.createAutoCorrectedDatePipe;
exports.mask = utils_dateCorrectionPipe.mask;
exports.IS_BROWSER = utils_nativeSupports.IS_BROWSER;
exports.SUPPORTS_INPUT_TYPE_DATE = utils_nativeSupports.SUPPORTS_INPUT_TYPE_DATE;
exports.isInputDateSupported = utils_nativeSupports.isInputDateSupported;
