// @alfalab/core-components/calendar-input/modern/utils.js
import { format, parse } from 'date-fns';

const DATE_FORMAT = 'dd.MM.yyyy';
const NATIVE_DATE_FORMAT = 'yyyy-MM-dd';
const DATE_MASK = [/\d/, /\d/, '.', /\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/];
const IS_BROWSER = typeof window !== 'undefined';
const SUPPORTS_INPUT_TYPE_DATE = IS_BROWSER && isInputDateSupported();
const isCompleteDateInput = (input) => input.length === DATE_MASK.length;
const formatDate = (date, dateFormat = DATE_FORMAT) => format(date, dateFormat);
const parseDateString = (value, dateFormat = DATE_FORMAT) => parse(value, dateFormat, new Date());
/**
 * Возвращает `true`, если поддерживается `input[type="date"]`
 */
function isInputDateSupported() {
    const input = document.createElement('input');
    const value = 'a';
    input.setAttribute('type', 'date');
    input.setAttribute('value', value);
    return input.value !== value;
}

export { DATE_FORMAT, DATE_MASK, IS_BROWSER, NATIVE_DATE_FORMAT, SUPPORTS_INPUT_TYPE_DATE, formatDate, isCompleteDateInput, isInputDateSupported, parseDateString };
