// @alfalab/core-components/tabs/esm/index.js
import './tslib.es6-44aaeed8.js';
import 'react';
import '../../keyboard-focusable/esm';
import 'classnames';
import 'compute-scroll-into-view';
export { ScrollableContainer } from './components/scrollable-container/Component.js';
export { useTabs } from './useTabs.js';
import './components/primary-tablist/Component.js';
import './index.module-c9e59889.js';
import './components/primary-tablist/Component.desktop.js';
import '../../tag/esm';
import './components/secondary-tablist/Component.js';
import './index.module-f5e5a0a0.js';
import './components/secondary-tablist/Component.desktop.js';
export { Tab } from './components/tab/Component.js';
import './components/tabs/Component.js';
import '@alfalab/hooks';
import './components/primary-tablist/Component.mobile.js';
export { PrimaryTabListResponsive } from './components/primary-tablist/Component.responsive.js';
import './components/secondary-tablist/Component.mobile.js';
export { SecondaryTabListResponsive } from './components/secondary-tablist/Component.responsive.js';
export { TabsResponsive } from './components/tabs/Component.responsive.js';
