// @alfalab/core-components/custom-button/modern/Component.js
import React from 'react';
import cn from 'classnames';
import { Button } from '../../button/modern';

var styles = {"customButton":"custom-button__customButton_e848k","white":"custom-button__white_e848k","black":"custom-button__black_e848k","customLoading":"custom-button__customLoading_e848k"};
require('./index.css');

const DEFAULT_BUTTON_COLOR = '#FF45C3';
const DEFAULT_CONTENT_COLOR = 'white';
const CustomButton = React.forwardRef(({ children, className, loading, backgroundColor = DEFAULT_BUTTON_COLOR, contentColor = DEFAULT_CONTENT_COLOR, ...restProps }, ref) => {
    const buttonProps = {
        style: { background: backgroundColor },
        ...restProps,
    };
    const buttonClassName = cn(styles.customButton, className, styles[contentColor], {
        [styles.customLoading]: loading,
    });
    return (React.createElement(Button, Object.assign({}, buttonProps, { view: 'primary', ref: ref, className: buttonClassName, loading: loading }), children));
});
/**
 * Для отображения в сторибуке
 */
CustomButton.defaultProps = {
    size: 'm',
    block: false,
    loading: false,
    nowrap: false,
};

export { CustomButton };
