// @alfalab/core-components/bottom-sheet/cssm/component.js
Object.defineProperty(exports, '__esModule', { value: true });

var components_swipeableBackdrop_Component = require('./Component-24200d96.js');
var React = require('react');
var cn = require('classnames');
var reactSwipeable = require('react-swipeable');
var coreComponentsBaseModal = require('../../base-modal/cssm');
var coreComponentsTypography = require('../../typography/cssm');
require('./components/footer/index.module.css');
var components_footer_Component = require('./components/footer/Component.js');
require('../../backdrop/cssm');
var styles = require('./index.module.css');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cn__default = /*#__PURE__*/_interopDefaultLegacy(cn);
var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

var TIMEOUT = 300;
var SWIPE_CLOSE_VELOCITY = 0.4;
var MIN_BACKDROP_OPACITY = 0.2;
var CLOSE_OFFSET = 0.2;
var BottomSheet = React.forwardRef(function (_a, ref) {
    var _b, _c;
    var open = _a.open, title = _a.title, actionButton = _a.actionButton, contentClassName = _a.contentClassName, className = _a.className, children = _a.children, zIndex = _a.zIndex, _d = _a.transitionProps, transitionProps = _d === void 0 ? {} : _d, dataTestId = _a.dataTestId, _e = _a.desktopSwipeable, trackMouse = _e === void 0 ? false : _e, onClose = _a.onClose;
    var _f = React.useState(0), sheetOffset = _f[0], setSheetOffset = _f[1];
    var _g = React.useState(1), backdropOpacity = _g[0], setBackdropOpacity = _g[1];
    var _h = React.useState(false), scrollLocked = _h[0], setScrollLocked = _h[1];
    var sheetHeight = React.useRef(0);
    var scrollableContainer = React.useRef(null);
    var scrollableContainerScrollValue = React.useRef(0);
    var getBackdropOpacity = function (offset) {
        if (sheetHeight.current === 0)
            return MIN_BACKDROP_OPACITY;
        var opacity = 1 - (1 - MIN_BACKDROP_OPACITY) * (offset / sheetHeight.current);
        return Number(opacity.toFixed(2));
    };
    var getSheetOffset = function (deltaY) {
        var offset = deltaY > 0 ? 0 : -deltaY;
        offset -= scrollableContainerScrollValue.current;
        return Math.floor(Math.max(0, offset));
    };
    /**
     * Если контент внутри шторки скроллится - то шторка не должна свайпаться
     */
    var shouldSkipSwiping = function () {
        if (!scrollableContainer.current) {
            return false;
        }
        if (!scrollableContainerScrollValue.current) {
            scrollableContainerScrollValue.current = Math.floor(scrollableContainer.current.scrollTop);
        }
        return scrollableContainer.current.scrollTop > 0;
    };
    var handleBackdropSwipedDown = function (_a) {
        var velocity = _a.velocity;
        if (velocity > SWIPE_CLOSE_VELOCITY) {
            onClose();
        }
    };
    var handleSheetSwipedDown = function (_a) {
        var velocity = _a.velocity;
        if (shouldSkipSwiping()) {
            return;
        }
        var shouldClose = sheetOffset > sheetHeight.current * CLOSE_OFFSET || velocity > SWIPE_CLOSE_VELOCITY;
        if (shouldClose) {
            onClose();
        }
        else {
            setSheetOffset(0);
            setBackdropOpacity(1);
        }
    };
    var handleSheetSwiped = function () {
        setScrollLocked(false);
        scrollableContainerScrollValue.current = 0;
    };
    var handleSheetSwiping = function (_a) {
        var deltaY = _a.deltaY;
        if (shouldSkipSwiping()) {
            return;
        }
        var offset = getSheetOffset(deltaY);
        var opacity = getBackdropOpacity(offset);
        setSheetOffset(offset);
        setBackdropOpacity(opacity);
        /**
         * Если шторка начинает свайпаться, то блокируем скролл внутри нее
         */
        if (offset > 0) {
            setScrollLocked(true);
        }
    };
    var backdropSwipeablehandlers = reactSwipeable.useSwipeable({
        onSwipedDown: handleBackdropSwipedDown,
        delta: 100,
        trackMouse: trackMouse,
    });
    var sheetSwipeablehandlers = reactSwipeable.useSwipeable({
        onSwiping: handleSheetSwiping,
        onSwipedDown: handleSheetSwipedDown,
        onSwiped: handleSheetSwiped,
        delta: 5,
        trackMouse: trackMouse,
    });
    var handleExited = React.useCallback(function (node) {
        setBackdropOpacity(1);
        if (transitionProps.onExited) {
            transitionProps.onExited(node);
        }
    }, [transitionProps]);
    var handleEntered = React.useCallback(function (node, isAppearing) {
        if (!sheetHeight.current) {
            sheetHeight.current = node.getBoundingClientRect().height;
        }
        setBackdropOpacity(1);
        if (transitionProps.onEntered) {
            transitionProps.onEntered(node, isAppearing);
        }
    }, [transitionProps]);
    React.useEffect(function () {
        if (!open) {
            setSheetOffset(0);
        }
    }, [open]);
    var getSwipeStyles = function () { return ({
        transform: sheetOffset ? "translateY(" + sheetOffset + "px)" : '',
    }); };
    return (React__default['default'].createElement(coreComponentsBaseModal.BaseModal, { open: open, ref: ref, dataTestId: dataTestId, zIndex: zIndex, onClose: onClose, scrollHandler: scrollableContainer, Backdrop: components_swipeableBackdrop_Component.SwipeableBackdrop, backdropProps: {
            opacity: backdropOpacity,
            handlers: backdropSwipeablehandlers,
            opacityTimeout: TIMEOUT,
        }, className: styles__default['default'].modal, transitionProps: components_swipeableBackdrop_Component.__assign(components_swipeableBackdrop_Component.__assign({ appear: true, timeout: TIMEOUT, classNames: styles__default['default'] }, transitionProps), { onExited: handleExited, onEntered: handleEntered }) },
        React__default['default'].createElement("div", components_swipeableBackdrop_Component.__assign({ className: cn__default['default'](styles__default['default'].component, className, (_b = {},
                _b[styles__default['default'].withTransition] = !sheetOffset,
                _b)), style: getSwipeStyles() }, sheetSwipeablehandlers),
            React__default['default'].createElement("div", { className: styles__default['default'].marker }),
            React__default['default'].createElement("div", { className: cn__default['default'](styles__default['default'].scrollableContainer, (_c = {},
                    _c[styles__default['default'].scrollLocked] = scrollLocked,
                    _c[styles__default['default'].withPadding] = !actionButton,
                    _c)), ref: scrollableContainer },
                title && (React__default['default'].createElement(coreComponentsTypography.Typography.Title, { view: 'small', font: 'system', tag: 'h2', className: styles__default['default'].title, color: 'primary' }, title)),
                React__default['default'].createElement("div", { className: cn__default['default'](styles__default['default'].content, contentClassName) }, children),
                actionButton && React__default['default'].createElement(components_footer_Component.Footer, null, actionButton)))));
});

exports.BottomSheet = BottomSheet;
exports.CLOSE_OFFSET = CLOSE_OFFSET;
