// @alfalab/core-components/toast/cssm/component.js
Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var mergeRefs = require('react-merge-refs');
var reactTransitionGroup = require('react-transition-group');
var cn = require('classnames');
var hooks = require('@alfalab/hooks');
var coreComponentsToastPlate = require('../../toast-plate/cssm');
var coreComponentsPopover = require('../../popover/cssm');
var coreComponentsPortal = require('../../portal/cssm');
var coreComponentsStack = require('../../stack/cssm');
var styles = require('./index.module.css');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var mergeRefs__default = /*#__PURE__*/_interopDefaultLegacy(mergeRefs);
var cn__default = /*#__PURE__*/_interopDefaultLegacy(cn);
var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign = function () {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
function __rest(s, e) {
    var t = {};
    for (var p in s)
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

var CSS_TRANSITION_CLASS_NAMES = {
    enter: styles__default['default'].enter,
    enterActive: styles__default['default'].enterActive,
    exit: styles__default['default'].exit,
    exitActive: styles__default['default'].exitActive,
};
var DefaultToastPlate = React.forwardRef(function (props, ref) { return (React__default['default'].createElement(coreComponentsToastPlate.ToastPlate, __assign({ ref: ref }, props))); });
var Toast = React.forwardRef(function (_a, ref) {
    var _b;
    var anchorElement = _a.anchorElement, position = _a.position, offset = _a.offset, open = _a.open, _c = _a.autoCloseDelay, autoCloseDelay = _c === void 0 ? 3000 : _c, className = _a.className, bottomOffset = _a.bottomOffset, _d = _a.style, style = _d === void 0 ? {} : _d, block = _a.block, _e = _a.zIndex, zIndex = _e === void 0 ? coreComponentsStack.stackingOrder.TOAST : _e, _f = _a.ToastPlate, ToastPlate = _f === void 0 ? DefaultToastPlate : _f, onMouseEnter = _a.onMouseEnter, onMouseLeave = _a.onMouseLeave, onTouchStart = _a.onTouchStart, onClose = _a.onClose, getPortalContainer = _a.getPortalContainer, restProps = __rest(_a, ["anchorElement", "position", "offset", "open", "autoCloseDelay", "className", "bottomOffset", "style", "block", "zIndex", "ToastPlate", "onMouseEnter", "onMouseLeave", "onTouchStart", "onClose", "getPortalContainer"]);
    var plateRef = React.useRef(null);
    var timerId = React.useRef(0);
    var prevOpen = hooks.usePrevious(open);
    var startTimer = React.useCallback(function () {
        clearTimeout(timerId.current);
        timerId.current = window.setTimeout(onClose, autoCloseDelay);
    }, [autoCloseDelay, onClose]);
    var stopTimer = React.useCallback(function () {
        clearTimeout(timerId.current);
    }, []);
    var handleMouseEnter = React.useCallback(function (event) {
        stopTimer();
        if (onMouseEnter) {
            onMouseEnter(event);
        }
    }, [onMouseEnter, stopTimer]);
    var handleMouseLeave = React.useCallback(function (event) {
        startTimer();
        if (onMouseLeave) {
            onMouseLeave(event);
        }
    }, [onMouseLeave, startTimer]);
    var handleTouchStart = React.useCallback(function (event) {
        stopTimer();
        if (onTouchStart) {
            onTouchStart(event);
        }
    }, [onTouchStart, stopTimer]);
    var handleClickOutside = React.useCallback(function () {
        onClose();
        stopTimer();
    }, [onClose, stopTimer]);
    hooks.useClickOutside(plateRef, handleClickOutside);
    React.useEffect(function () {
        if (open !== prevOpen && open) {
            startTimer();
        }
    }, [open, prevOpen, startTimer, stopTimer]);
    var props = {
        block: block,
        onClose: onClose,
        onMouseEnter: handleMouseEnter,
        onMouseLeave: handleMouseLeave,
        onTouchStart: handleTouchStart,
        ref: mergeRefs__default['default']([ref, plateRef]),
    };
    if (anchorElement) {
        return (React__default['default'].createElement(coreComponentsPopover.Popover, { open: open, anchorElement: anchorElement, position: position, offset: offset, popperClassName: styles__default['default'].popoverInner, className: cn__default['default']((_b = {}, _b[styles__default['default'].block] = block, _b)), transition: { timeout: 150 }, getPortalContainer: getPortalContainer, zIndex: zIndex },
            React__default['default'].createElement(ToastPlate, __assign({}, restProps, { style: style, className: className }, props))));
    }
    return (React__default['default'].createElement(coreComponentsStack.Stack, { value: zIndex }, function (computedZIndex) { return (React__default['default'].createElement(coreComponentsPortal.Portal, { getPortalContainer: getPortalContainer },
        React__default['default'].createElement(reactTransitionGroup.CSSTransition, { unmountOnExit: true, in: open, timeout: 150, classNames: CSS_TRANSITION_CLASS_NAMES },
            React__default['default'].createElement(ToastPlate, __assign({}, restProps, { className: cn__default['default'](styles__default['default'].fixed, styles__default['default'].toastPlate, className), style: __assign(__assign({}, style), { bottom: bottomOffset && bottomOffset + "px", zIndex: computedZIndex }) }, props))))); }));
});

exports.Toast = Toast;
