// @alfalab/core-components/select/modern/index.js
import 'react';
import 'classnames';
import '@alfalab/hooks';
import '@alfalab/icons-glyph/ChevronDownMIcon';
import 'react-merge-refs';
import 'resize-observer';
import 'downshift';
import 'react-virtual';
import '../../form-control/modern';
export { isGroup, isOptionShape, joinOptions, lastIndexOf, processOptions, usePrevious, useVisibleOptions } from './utils.js';
export { Field } from './components/field/Component.js';
export { Arrow } from './components/arrow/Component.js';
export { Optgroup } from './components/optgroup/Component.js';
export { OptionsList } from './components/options-list/Component.js';
import '../../checkbox/modern';
import './components/checkmark/Component.js';
export { Option } from './components/option/Component.js';
import '../../popover/modern';
import './components/native-select/Component.js';
export { BaseSelect } from './components/base-select/Component.js';
export { Select } from './Component.js';
import '../../skeleton/modern';
import './intersection-observer-de5aa0a1.js';
import '../../button/modern';
export { S as SELECT_ALL_KEY, u as useSelectWithApply } from './hook-817cbffc.js';
export { VirtualOptionsList } from './components/virtual-options-list/Component.js';
export { useSelectWithLoading } from './presets/useSelectWithLoading/hook.js';
export { useLazyLoading } from './presets/useLazyLoading/hook.js';
