// @alfalab/core-components/gallery/modern/index.js
import 'react';
import 'classnames';
import '@alfalab/hooks';
import '@alfalab/icons-glyph/PointerDownMIcon';
import '@alfalab/icons-glyph/ArrowsOutwardMIcon';
import '@alfalab/icons-glyph/ArrowsInwardMIcon';
import '@alfalab/icons-glyph/CrossMIcon';
import 'swiper';
import 'swiper/react';
import 'element-closest';
import 'swiper/swiper.min.css';
import '@alfalab/icons-glyph/ChevronBackHeavyMIcon';
import '@alfalab/icons-glyph/ChevronForwardHeavyMIcon';
import '../../base-modal/modern';
import './context.js';
import './components/image-preview/Component.js';
export { PLACEHOLDER_HEIGHT, PLACEHOLDER_WIDTH, getImageAlt, getImageKey, isSmallImage } from './utils/utils.js';
export { TestIds } from './utils/constants.js';
import './components/navigation-bar/Component.js';
export { splitFilename } from './utils/split-filename.js';
import '../../typography/modern';
import './components/header-info-block/Component.js';
import '../../icon-button/modern';
import '../../tooltip/modern';
import './components/header/buttons.js';
import './components/header/Component.js';
import './slide-d0e4396e.js';
import './components/image-viewer/component.js';
export { Gallery } from './Component.js';
