// @alfalab/core-components/calendar-input/cssm/utils.js
Object.defineProperty(exports, '__esModule', { value: true });

var dateFns = require('date-fns');

var DATE_FORMAT = 'dd.MM.yyyy';
var NATIVE_DATE_FORMAT = 'yyyy-MM-dd';
var DATE_MASK = [/\d/, /\d/, '.', /\d/, /\d/, '.', /\d/, /\d/, /\d/, /\d/];
var IS_BROWSER = typeof window !== 'undefined';
var SUPPORTS_INPUT_TYPE_DATE = IS_BROWSER && isInputDateSupported();
var isCompleteDateInput = function (input) { return input.length === DATE_MASK.length; };
var formatDate = function (date, dateFormat) {
    if (dateFormat === void 0) { dateFormat = DATE_FORMAT; }
    return dateFns.format(date, dateFormat);
};
var parseDateString = function (value, dateFormat) {
    if (dateFormat === void 0) { dateFormat = DATE_FORMAT; }
    return dateFns.parse(value, dateFormat, new Date());
};
/**
 * Возвращает `true`, если поддерживается `input[type="date"]`
 */
function isInputDateSupported() {
    var input = document.createElement('input');
    var value = 'a';
    input.setAttribute('type', 'date');
    input.setAttribute('value', value);
    return input.value !== value;
}

exports.DATE_FORMAT = DATE_FORMAT;
exports.DATE_MASK = DATE_MASK;
exports.IS_BROWSER = IS_BROWSER;
exports.NATIVE_DATE_FORMAT = NATIVE_DATE_FORMAT;
exports.SUPPORTS_INPUT_TYPE_DATE = SUPPORTS_INPUT_TYPE_DATE;
exports.formatDate = formatDate;
exports.isCompleteDateInput = isCompleteDateInput;
exports.isInputDateSupported = isInputDateSupported;
exports.parseDateString = parseDateString;
