// @alfalab/core-components/link/cssm/Component.js
Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var cn = require('classnames');
var mergeRefs = require('react-merge-refs');
var hooks = require('@alfalab/hooks');
var styles = require('./index.module.css');
var defaultColors = require('./default.module.css');
var invertedColors = require('./inverted.module.css');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cn__default = /*#__PURE__*/_interopDefaultLegacy(cn);
var mergeRefs__default = /*#__PURE__*/_interopDefaultLegacy(mergeRefs);
var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);
var defaultColors__default = /*#__PURE__*/_interopDefaultLegacy(defaultColors);
var invertedColors__default = /*#__PURE__*/_interopDefaultLegacy(invertedColors);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign = function () {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
function __rest(s, e) {
    var t = {};
    for (var p in s)
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

var colorStyles = {
    default: defaultColors__default['default'],
    inverted: invertedColors__default['default'],
};
var Link = React.forwardRef(function (_a, ref) {
    var _b, _c;
    var _d = _a.view, view = _d === void 0 ? 'primary' : _d, _e = _a.pseudo, pseudo = _e === void 0 ? false : _e, leftAddons = _a.leftAddons, rightAddons = _a.rightAddons, className = _a.className, dataTestId = _a.dataTestId, children = _a.children, _f = _a.colors, colors = _f === void 0 ? 'default' : _f, href = _a.href, _g = _a.Component, Component = _g === void 0 ? 'a' : _g, restProps = __rest(_a, ["view", "pseudo", "leftAddons", "rightAddons", "className", "dataTestId", "children", "colors", "href", "Component"]);
    var linkRef = React.useRef(null);
    var focused = hooks.useFocus(linkRef, 'keyboard')[0];
    var componentProps = (_b = {
            className: cn__default['default'](styles__default['default'].component, colorStyles[colors][view], (_c = {},
                _c[styles__default['default'].pseudo] = pseudo,
                _c[styles__default['default'].focused] = focused,
                _c[styles__default['default'].withAddons] = leftAddons || rightAddons,
                _c), className),
            'data-test-id': dataTestId,
            rel: restProps.target === '_blank' ? 'noreferrer noopener' : undefined
        },
        // Для совместимости с react-router-dom, меняем href на to
        _b[typeof Component === 'string' ? 'href' : 'to'] = href,
        _b);
    return (React__default['default'].createElement(Component, __assign({}, componentProps, restProps, { ref: mergeRefs__default['default']([linkRef, ref]) }), leftAddons || rightAddons ? (React__default['default'].createElement(React__default['default'].Fragment, null,
        leftAddons && React__default['default'].createElement("span", { className: styles__default['default'].addons }, leftAddons),
        children && (React__default['default'].createElement("span", null,
            React__default['default'].createElement("span", { className: styles__default['default'].text }, children))),
        rightAddons && React__default['default'].createElement("span", { className: styles__default['default'].addons }, rightAddons))) : (React__default['default'].createElement("span", { className: styles__default['default'].text }, children))));
});
/**
 * Для отображения в сторибуке
 */
Link.defaultProps = {
    view: 'primary',
    pseudo: false,
};

exports.Link = Link;
