// @alfalab/core-components/notification/Component.js
Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var cn = require('classnames');
var mergeRefs = require('react-merge-refs');
var reactSwipeable = require('react-swipeable');
var elementClosest = require('element-closest');
var coreComponentsPortal = require('../portal');
var coreComponentsToastPlate = require('../toast-plate');
var coreComponentsStack = require('../stack');
var utils_index = require('./utils/index.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cn__default = /*#__PURE__*/_interopDefaultLegacy(cn);
var mergeRefs__default = /*#__PURE__*/_interopDefaultLegacy(mergeRefs);
var elementClosest__default = /*#__PURE__*/_interopDefaultLegacy(elementClosest);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign = function () {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
function __rest(s, e) {
    var t = {};
    for (var p in s)
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

var styles = {"notificationComponent":"notification__notificationComponent_1hvmu","isVisible":"notification__isVisible_1hvmu","isClosing":"notification__isClosing_1hvmu","toastContent":"notification__toastContent_1hvmu"};
require('./index.css');

var notificationClassNameSelector = "." + styles.notificationComponent;
var Notification = React.forwardRef(function (_a, ref) {
    var className = _a.className, children = _a.children, visible = _a.visible, _b = _a.offset, offset = _b === void 0 ? 108 : _b, _c = _a.hasCloser, hasCloser = _c === void 0 ? true : _c, _d = _a.autoCloseDelay, autoCloseDelay = _d === void 0 ? 5000 : _d, _e = _a.usePortal, usePortal = _e === void 0 ? true : _e, _f = _a.zIndex, zIndex = _f === void 0 ? coreComponentsStack.stackingOrder.TOAST : _f, style = _a.style, onClose = _a.onClose, onCloseTimeout = _a.onCloseTimeout, onMouseEnter = _a.onMouseEnter, onMouseLeave = _a.onMouseLeave, onClickOutside = _a.onClickOutside, restProps = __rest(_a, ["className", "children", "visible", "offset", "hasCloser", "autoCloseDelay", "usePortal", "zIndex", "style", "onClose", "onCloseTimeout", "onMouseEnter", "onMouseLeave", "onClickOutside"]);
    var notificationRef = React.useRef(null);
    var autoCloseTimeoutRef = React.useRef(0);
    var closeTimeoutRef = React.useRef(0);
    var _g = React.useState(false), isClosing = _g[0], setIsClosing = _g[1];
    var startAutoCloseTimer = React.useCallback(function () {
        if (autoCloseDelay !== null) {
            autoCloseTimeoutRef.current = window.setTimeout(function () {
                if (onCloseTimeout) {
                    onCloseTimeout();
                }
            }, autoCloseDelay);
        }
    }, [autoCloseDelay, onCloseTimeout]);
    var stopAutoCloseTimer = React.useCallback(function () {
        clearTimeout(autoCloseTimeoutRef.current);
    }, []);
    React.useEffect(function () { return function () {
        clearTimeout(closeTimeoutRef.current);
    }; }, []);
    React.useEffect(function () {
        elementClosest__default['default'](window);
    }, []);
    React.useEffect(function () {
        if (visible) {
            startAutoCloseTimer();
        }
        return function () {
            stopAutoCloseTimer();
        };
    }, [startAutoCloseTimer, stopAutoCloseTimer, visible]);
    var handleMouseEnter = React.useCallback(function (event) {
        stopAutoCloseTimer();
        if (onMouseEnter) {
            onMouseEnter(event);
        }
    }, [onMouseEnter, stopAutoCloseTimer]);
    var handleMouseLeave = React.useCallback(function (event) {
        stopAutoCloseTimer();
        startAutoCloseTimer();
        if (onMouseLeave) {
            onMouseLeave(event);
        }
    }, [onMouseLeave, startAutoCloseTimer, stopAutoCloseTimer]);
    var handleOutsideClick = React.useCallback(function (event) {
        var isTargetNotification = !!event.target.closest(notificationClassNameSelector);
        /*
         *  проверка isTargetNotification нужна для предотвращения срабатывания handleOutsideClick
         *  при клике на другие нотификации, если их несколько на странице
         */
        if (onClickOutside && visible && !isTargetNotification) {
            onClickOutside(event);
        }
    }, [onClickOutside, visible]);
    utils_index.useClickOutside(notificationRef, handleOutsideClick);
    var swipeableHandlers = reactSwipeable.useSwipeable({
        onSwiped: function (_a) {
            var dir = _a.dir;
            if (onClose && ['Left', 'Right', 'Up'].includes(dir)) {
                setIsClosing(true);
                closeTimeoutRef.current = window.setTimeout(function () {
                    setIsClosing(false);
                    onClose();
                }, 100);
            }
        },
        delta: 100,
    });
    var Wrapper = usePortal ? coreComponentsPortal.Portal : React.Fragment;
    return (React__default['default'].createElement(coreComponentsStack.Stack, { value: zIndex }, function (computedZIndex) {
        var _a;
        return (React__default['default'].createElement(Wrapper, null,
            React__default['default'].createElement("div", __assign({}, swipeableHandlers),
                React__default['default'].createElement(coreComponentsToastPlate.ToastPlate, __assign({ className: cn__default['default'](styles.notificationComponent, (_a = {},
                        _a[styles.isVisible] = visible,
                        _a[styles.isClosing] = isClosing,
                        _a), className), contentClassName: styles.toastContent, style: __assign({ top: offset, zIndex: computedZIndex }, style), onMouseEnter: handleMouseEnter, onMouseLeave: handleMouseLeave, ref: mergeRefs__default['default']([ref, notificationRef]), role: visible ? 'alert' : undefined, hasCloser: hasCloser, onClose: onClose }, restProps), children))));
    }));
});
/**
 * Для отображения в сторибуке
 */
Notification.defaultProps = {
    autoCloseDelay: 5000,
    offset: 108,
    hasCloser: true,
    usePortal: true,
};

exports.Notification = Notification;
