// @alfalab/core-components/calendar-range/Component.js
Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var cn = require('classnames');
var dateFns = require('date-fns');
var coreComponentsCalendar = require('../calendar');
var coreComponentsCalendarInput = require('../calendar-input');
var utils = require('./utils.js');
var useCalendarMonthes = require('./useCalendarMonthes.js');
var useCalendarMaxMinDates = require('./useCalendarMaxMinDates.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cn__default = /*#__PURE__*/_interopDefaultLegacy(cn);

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign = function () {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};

var styles = {"component":"calendar-range__component_ft44u","divider":"calendar-range__divider_ft44u"};
require('./index.css');

var CalendarRange = function (_a) {
    var className = _a.className, _b = _a.defaultMonth, defaultMonth = _b === void 0 ? dateFns.startOfMonth(new Date()).getTime() : _b, minDate = _a.minDate, maxDate = _a.maxDate, valueFrom = _a.valueFrom, valueTo = _a.valueTo, onDateFromChange = _a.onDateFromChange, onDateToChange = _a.onDateToChange, _c = _a.inputFromProps, inputFromProps = _c === void 0 ? {} : _c, _d = _a.inputToProps, inputToProps = _d === void 0 ? {} : _d, _e = _a.calendarPosition, calendarPosition = _e === void 0 ? 'static' : _e, dataTestId = _a.dataTestId;
    var uncontrolled = valueFrom === undefined && valueTo === undefined;
    var isPopover = calendarPosition === 'popover';
    var period = coreComponentsCalendar.usePeriod({
        initialSelectedFrom: valueFrom ? coreComponentsCalendarInput.parseDateString(valueFrom).getTime() : undefined,
        initialSelectedTo: valueTo ? coreComponentsCalendarInput.parseDateString(valueTo).getTime() : undefined,
    });
    var setStart = period.setStart, setEnd = period.setEnd, resetPeriod = period.resetPeriod;
    var selectedFrom = period.selectedFrom, selectedTo = period.selectedTo;
    if (!coreComponentsCalendar.dateInLimits(selectedFrom, minDate, maxDate))
        selectedFrom = undefined;
    if (!coreComponentsCalendar.dateInLimits(selectedTo, minDate, maxDate))
        selectedTo = undefined;
    var _f = React.useState(utils.initialValueState), stateFrom = _f[0], setStateFrom = _f[1];
    var _g = React.useState(utils.initialValueState), stateTo = _g[0], setStateTo = _g[1];
    var inputValueFrom = utils.getCorrectValueState(stateFrom, valueFrom, minDate, maxDate);
    var inputValueTo = utils.getCorrectValueState(stateTo, valueTo, minDate, maxDate);
    var handleStateFromChange = React.useCallback(function (newFromState) {
        if (uncontrolled)
            setStateFrom(newFromState);
        if (onDateFromChange)
            onDateFromChange(newFromState);
    }, [onDateFromChange, uncontrolled]);
    var handleStateToChange = React.useCallback(function (newToState) {
        if (uncontrolled)
            setStateTo(newToState);
        if (onDateToChange)
            onDateToChange(newToState);
    }, [onDateToChange, uncontrolled]);
    var _h = useCalendarMonthes.useCalendarMonthes({
        inputValueFrom: inputValueFrom,
        inputValueTo: inputValueTo,
        defaultMonth: defaultMonth,
        isPopover: isPopover,
    }), monthFrom = _h.monthFrom, monthTo = _h.monthTo, handleMonthFromChange = _h.handleMonthFromChange, handleMonthToChange = _h.handleMonthToChange;
    var handleInputFromChange = React.useCallback(function (_, _a) {
        var value = _a.value, date = _a.date;
        if (value === '') {
            setStart(undefined);
            handleStateFromChange(utils.initialValueState);
        }
        if (!coreComponentsCalendarInput.isCompleteDateInput(value)) {
            return;
        }
        if (coreComponentsCalendar.dateInLimits(date, minDate, maxDate)) {
            setStart(date.getTime());
            handleMonthFromChange(dateFns.startOfMonth(date).getTime());
            handleStateFromChange({ date: date.getTime(), value: value });
        }
        else {
            setStart(undefined);
            handleStateFromChange({ date: null, value: value });
        }
    }, [handleMonthFromChange, handleStateFromChange, maxDate, minDate, setStart]);
    var handleInputToChange = React.useCallback(function (_, _a) {
        var value = _a.value, date = _a.date;
        if (value === '') {
            setEnd(undefined);
            handleStateToChange(utils.initialValueState);
        }
        if (!coreComponentsCalendarInput.isCompleteDateInput(value)) {
            return;
        }
        if (coreComponentsCalendar.dateInLimits(date, minDate, maxDate)) {
            setEnd(date.getTime());
            handleMonthToChange(dateFns.subMonths(dateFns.startOfMonth(date), 1).getTime());
            handleStateToChange({ date: date.getTime(), value: value });
        }
        else {
            setEnd(undefined);
            handleStateToChange({ date: null, value: value });
        }
    }, [handleMonthToChange, handleStateToChange, maxDate, minDate, setEnd]);
    var handleCalendarChange = React.useCallback(function (date) {
        if (!inputValueFrom.date) {
            setStart(date);
            handleStateFromChange({ date: date, value: coreComponentsCalendarInput.formatDate(date) });
            return;
        }
        if (date < inputValueFrom.date) {
            resetPeriod();
            setStart(date);
            handleStateToChange(utils.initialValueState);
            handleStateFromChange({ date: date, value: coreComponentsCalendarInput.formatDate(date) });
            return;
        }
        if (date === inputValueFrom.date && date === inputValueTo.date) {
            resetPeriod();
            handleStateFromChange(utils.initialValueState);
            handleStateToChange(utils.initialValueState);
            return;
        }
        handleStateToChange({ date: date, value: coreComponentsCalendarInput.formatDate(date) });
        setEnd(date);
    }, [
        inputValueFrom.date,
        inputValueTo.date,
        handleStateToChange,
        setEnd,
        setStart,
        handleStateFromChange,
        resetPeriod,
    ]);
    var handleFromCalendarChange = React.useCallback(function (date) {
        if (!isPopover) {
            handleCalendarChange(date);
            return;
        }
        setStart(date);
        handleStateFromChange({ date: date, value: coreComponentsCalendarInput.formatDate(date) });
    }, [handleCalendarChange, handleStateFromChange, isPopover, setStart]);
    var handleToCalendarChange = React.useCallback(function (date) {
        if (!isPopover) {
            handleCalendarChange(date);
            return;
        }
        handleStateToChange({ date: date, value: coreComponentsCalendarInput.formatDate(date) });
        setEnd(date);
    }, [handleCalendarChange, handleStateToChange, isPopover, setEnd]);
    var _j = React.useState(false), nextMonthHighlighted = _j[0], setNextMonthHighlighted = _j[1];
    var handleCalendarToMouseOver = React.useCallback(function (event) {
        var target = event.target;
        var dayHighlighted = utils.isDayButton(target) || utils.isDayButton(target.parentElement);
        if (nextMonthHighlighted && !dayHighlighted)
            setNextMonthHighlighted(false);
        if (!nextMonthHighlighted && dayHighlighted)
            setNextMonthHighlighted(true);
    }, [nextMonthHighlighted]);
    var selectorView = isPopover ? 'full' : 'month-only';
    var calendarSelectedTo = selectedTo || (nextMonthHighlighted ? monthTo : undefined);
    var maxMinDates = useCalendarMaxMinDates.useCalendarMaxMinDates({
        isPopover: isPopover,
        monthTo: monthTo,
        monthFrom: monthFrom,
        selectedFrom: selectedFrom,
        selectedTo: calendarSelectedTo,
        maxDate: maxDate,
        minDate: minDate,
    });
    return (React__default['default'].createElement("div", { className: cn__default['default'](styles.component, className), "data-test-id": dataTestId },
        React__default['default'].createElement(coreComponentsCalendarInput.CalendarInput, __assign({}, inputFromProps, { useAnchorWidth: false, calendarPosition: calendarPosition, onInputChange: handleInputFromChange, onCalendarChange: handleFromCalendarChange, value: inputValueFrom.value, minDate: maxMinDates.fromMinDate, maxDate: maxMinDates.fromMaxDate, calendarProps: __assign(__assign({}, inputFromProps.calendarProps), { month: monthFrom, onMonthChange: handleMonthFromChange, selectorView: selectorView,
                selectedFrom: selectedFrom, selectedTo: calendarSelectedTo }) })),
        React__default['default'].createElement("span", { className: styles.divider }),
        React__default['default'].createElement("div", { onMouseOver: handleCalendarToMouseOver },
            React__default['default'].createElement(coreComponentsCalendarInput.CalendarInput, __assign({}, inputToProps, { useAnchorWidth: false, calendarPosition: calendarPosition, popoverPosition: 'bottom-end', onInputChange: handleInputToChange, onCalendarChange: handleToCalendarChange, value: inputValueTo.value, minDate: maxMinDates.toMinDate, maxDate: maxMinDates.toMaxDate, calendarProps: __assign(__assign({}, inputToProps.calendarProps), { month: monthTo, onMonthChange: handleMonthToChange, selectorView: selectorView,
                    selectedFrom: selectedFrom,
                    selectedTo: selectedTo }) })))));
};

exports.CalendarRange = CalendarRange;
