// @alfalab/core-components/chart/cssm/components/Legends/index.js
Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var cn = require('classnames');
var coreComponentsTypography = require('../../../../typography/cssm');
var icons_Circle = require('../../icons/Circle.js');
var icons_CircleLine = require('../../icons/CircleLine.js');
var icons_FilledCircle = require('../../icons/FilledCircle.js');
var icons_StrokeCircle = require('../../icons/StrokeCircle.js');
var styles = require('./index.module.css');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cn__default = /*#__PURE__*/_interopDefaultLegacy(cn);
var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

var icons = {
    circleLine: icons_CircleLine.CircleLineIcon,
    filledCircle: icons_FilledCircle.FilledCircleIcon,
    strokeCircle: icons_StrokeCircle.StrokeCircleIcon,
    circle: icons_Circle.CircleIcon,
};
var Legends = React__default['default'].forwardRef(function (_a, ref) {
    var legend = _a.legend, series = _a.series, id = _a.id, charts = _a.charts, toggleChart = _a.toggleChart;
    var style = {
        textAlign: legend.align || 'center',
        transform: "translateY(" + ((legend === null || legend === void 0 ? void 0 : legend.marginTop) ? legend.marginTop : 0) *
            (legend.verticalAlign === 'top' ? -1 : 1) + "px)",
    };
    return (React__default['default'].createElement("ul", { ref: ref, className: cn__default['default'](styles__default['default'].legendWrap), style: style }, series.map(function (item) {
        var _a, _b;
        if (item.hideLegend || item.hide)
            return null;
        var Icon = icons[item.icon] || icons_Circle.CircleIcon;
        return (React__default['default'].createElement("li", { role: 'presentation', key: id + "-" + item.properties.dataKey, className: cn__default['default'](styles__default['default'].legendItem, charts["" + item.properties.dataKey] ? '' : styles__default['default'].legendUnactive), onClick: function () { return toggleChart(item); } },
            React__default['default'].createElement("div", { className: cn__default['default'](styles__default['default'].legendContent) },
                Icon ? (React__default['default'].createElement("i", { className: cn__default['default'](styles__default['default'].legendIcon) },
                    React__default['default'].createElement(Icon, { fill: ((_a = item.properties) === null || _a === void 0 ? void 0 : _a.fill) || ((_b = item.properties) === null || _b === void 0 ? void 0 : _b.stroke) ||
                            '', height: legend.iconHeight || 16 }))) : null,
                React__default['default'].createElement(coreComponentsTypography.Typography.Text, { view: 'primary-medium', tag: 'span', className: cn__default['default'](styles__default['default'].legendValue) }, item.properties.name))));
    })));
});

exports.Legends = Legends;
