// @alfalab/core-components/select/hook-dbd120ed.js
var tslib_es6 = require('./tslib.es6-7a40dfc0.js');
var React = require('react');
var cn = require('classnames');
require('@alfalab/hooks');
require('../form-control');
var utils = require('./utils.js');
require('@alfalab/icons-glyph/ChevronDownMIcon');
var components_optionsList_Component = require('./components/options-list/Component.js');
require('../checkbox');
require('./components/base-select/Component.js');
require('./Component.js');
require('react-virtual');
require('../skeleton');
require('./intersection-observer-b603d85e.js');
var coreComponentsButton = require('../button');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var cn__default = /*#__PURE__*/_interopDefaultLegacy(cn);

var styles = {"footer":"select__footer_18ac7","withBorder":"select__withBorder_18ac7"};
require('./presets/useSelectWithApply/options-list-with-apply/index.css');

var OptionsListWithApply = React.forwardRef(function (_a, ref) {
    var _b;
    var toggleMenu = _a.toggleMenu, _c = _a.OptionsList, OptionsList = _c === void 0 ? components_optionsList_Component.OptionsList : _c, defaultGetOptionProps = _a.getOptionProps, _d = _a.showClear, showClear = _d === void 0 ? true : _d, _e = _a.selectedDraft, selectedDraft = _e === void 0 ? [] : _e, _f = _a.flatOptions, flatOptions = _f === void 0 ? [] : _f, _g = _a.onApply, onApply = _g === void 0 ? function () { return null; } : _g, _h = _a.onClear, onClear = _h === void 0 ? function () { return null; } : _h, _j = _a.onClose, onClose = _j === void 0 ? function () { return null; } : _j, _k = _a.visibleOptions, visibleOptions = _k === void 0 ? 5 : _k, restProps = tslib_es6.__rest(_a, ["toggleMenu", "OptionsList", "getOptionProps", "showClear", "selectedDraft", "flatOptions", "onApply", "onClear", "onClose", "visibleOptions"]);
    var footerRef = React.useRef(null);
    var getOptionProps = React.useCallback(function (option, index) {
        var optionProps = defaultGetOptionProps(option, index);
        var selected = option.key === SELECT_ALL_KEY
            ? selectedDraft.length === flatOptions.length - 1
            : selectedDraft.includes(option);
        return tslib_es6.__assign(tslib_es6.__assign({}, optionProps), { selected: selected });
    }, [defaultGetOptionProps, flatOptions.length, selectedDraft]);
    var handleApply = React.useCallback(function () {
        onApply();
        toggleMenu();
    }, [onApply, toggleMenu]);
    var handleClear = React.useCallback(function () {
        onClear();
        toggleMenu();
    }, [onClear, toggleMenu]);
    React.useEffect(function () {
        var activeElement = document.activeElement;
        setTimeout(function () {
            if (footerRef.current) {
                footerRef.current.focus();
            }
        }, 0);
        return function () {
            onClose();
            if (activeElement) {
                activeElement.focus();
            }
        };
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);
    return (React__default['default'].createElement(OptionsList, tslib_es6.__assign({}, restProps, { ref: ref, visibleOptions: visibleOptions, toggleMenu: toggleMenu, flatOptions: flatOptions, getOptionProps: getOptionProps, footer: React__default['default'].createElement("div", { 
            // eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex
            tabIndex: 0, className: cn__default['default'](styles.footer, (_b = {},
                _b[styles.withBorder] = visibleOptions && flatOptions.length > visibleOptions,
                _b)), ref: footerRef },
            React__default['default'].createElement(coreComponentsButton.Button, { size: 'xxs', view: 'primary', onClick: handleApply }, "\u041F\u0440\u0438\u043C\u0435\u043D\u0438\u0442\u044C"),
            showClear && (React__default['default'].createElement(coreComponentsButton.Button, { size: 'xxs', view: 'secondary', onClick: handleClear }, "\u0421\u0431\u0440\u043E\u0441\u0438\u0442\u044C"))) })));
});

var SELECT_ALL_KEY = 'select_all';
var selectAllOption = { key: SELECT_ALL_KEY, content: 'Выбрать все' };
function useSelectWithApply(_a) {
    var options = _a.options, selected = _a.selected, _b = _a.onChange, onChange = _b === void 0 ? function () { return null; } : _b, OptionsList = _a.OptionsList, _c = _a.showClear, showClear = _c === void 0 ? true : _c, _d = _a.showSelectAll, showSelectAll = _d === void 0 ? false : _d;
    var _e = React.useMemo(function () { return utils.processOptions(options, selected); }, [
        options,
        selected,
    ]), flatOptions = _e.flatOptions, selectedOptions = _e.selectedOptions;
    var _f = React.useState(selectedOptions), selectedDraft = _f[0], setSelectedDraft = _f[1];
    var selectedOptionsRef = React.useRef(selectedOptions);
    var handleApply = React.useCallback(function () {
        onChange({
            selected: selectedDraft[0],
            selectedMultiple: selectedDraft,
            initiator: null,
        });
    }, [onChange, selectedDraft]);
    var handleClear = React.useCallback(function () {
        setSelectedDraft([]);
        onChange({
            selected: null,
            selectedMultiple: [],
            initiator: null,
        });
    }, [onChange]);
    var handleChange = React.useCallback(function (_a) {
        var initiator = _a.initiator, restArgs = tslib_es6.__rest(_a, ["initiator"]);
        if (!initiator) {
            onChange(tslib_es6.__assign({ initiator: null }, restArgs));
            return;
        }
        var initiatorSelected = selectedDraft.includes(initiator) ||
            (initiator.key === SELECT_ALL_KEY && selectedDraft.length === flatOptions.length);
        if (initiator.key === SELECT_ALL_KEY) {
            setSelectedDraft(initiatorSelected ? [] : flatOptions);
        }
        else {
            setSelectedDraft(initiatorSelected
                ? selectedDraft.filter(function (o) { return o !== initiator; })
                : selectedDraft.concat(initiator));
        }
    }, [flatOptions, onChange, selectedDraft]);
    var handleClose = React.useCallback(function () {
        setSelectedDraft(selectedOptionsRef.current);
    }, []);
    React.useEffect(function () {
        setSelectedDraft(selectedOptions);
        selectedOptionsRef.current = selectedOptions;
    }, [selectedOptions]);
    var memoizedOptions = React.useMemo(function () { return (showSelectAll ? tslib_es6.__spreadArrays([selectAllOption], options) : options); }, [options, showSelectAll]);
    return {
        OptionsList: OptionsListWithApply,
        optionsListProps: {
            OptionsList: OptionsList,
            showClear: showClear && (selectedDraft.length > 0 || selectedOptions.length > 0),
            onClear: handleClear,
            onApply: handleApply,
            onClose: handleClose,
            selectedDraft: selectedDraft,
        },
        allowUnselect: true,
        multiple: true,
        options: memoizedOptions,
        onChange: handleChange,
        selected: selected,
    };
}

exports.OptionsListWithApply = OptionsListWithApply;
exports.SELECT_ALL_KEY = SELECT_ALL_KEY;
exports.useSelectWithApply = useSelectWithApply;
