// @alfalab/core-components/slider-input/esm/Component.js
import React, { forwardRef, useCallback, Fragment, isValidElement, cloneElement } from 'react';
import cn from 'classnames';
import { Slider } from '../../slider/esm';
import { Input } from '../../input/esm';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign = function () {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
function __rest(s, e) {
    var t = {};
    for (var p in s)
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

var styles = {"component":"slider-input__component_2h449","block":"slider-input__block_2h449","slider":"slider-input__slider_2h449","field":"slider-input__field_2h449","input":"slider-input__input_2h449","steps":"slider-input__steps_2h449","info":"slider-input__info_2h449","s":"slider-input__s_2h449","hasLabel":"slider-input__hasLabel_2h449","m":"slider-input__m_2h449","l":"slider-input__l_2h449","xl":"slider-input__xl_2h449","focused":"slider-input__focused_2h449","filled":"slider-input__filled_2h449"};
require('./index.css');

var SliderInput = forwardRef(function (_a, ref) {
    var _b;
    var className = _a.className, inputClassName = _a.inputClassName, sliderClassName = _a.sliderClassName, stepsClassName = _a.stepsClassName, focusedClassName = _a.focusedClassName, fieldClassName = _a.fieldClassName, _c = _a.value, value = _c === void 0 ? '' : _c, _d = _a.min, min = _d === void 0 ? 0 : _d, _e = _a.max, max = _e === void 0 ? 100 : _e, _f = _a.step, step = _f === void 0 ? 1 : _f, block = _a.block, _g = _a.sliderValue, sliderValue = _g === void 0 ? +value : _g, _h = _a.steps, steps = _h === void 0 ? [] : _h, _j = _a.size, size = _j === void 0 ? 's' : _j, label = _a.label, info = _a.info, disabled = _a.disabled, readOnly = _a.readOnly, onChange = _a.onChange, onInputChange = _a.onInputChange, onSliderChange = _a.onSliderChange, rightAddons = _a.rightAddons, _k = _a.Input, Input$1 = _k === void 0 ? Input : _k, _l = _a.customInputProps, customInputProps = _l === void 0 ? {} : _l, dataTestId = _a.dataTestId, error = _a.error, restProps = __rest(_a, ["className", "inputClassName", "sliderClassName", "stepsClassName", "focusedClassName", "fieldClassName", "value", "min", "max", "step", "block", "sliderValue", "steps", "size", "label", "info", "disabled", "readOnly", "onChange", "onInputChange", "onSliderChange", "rightAddons", "Input", "customInputProps", "dataTestId", "error"]);
    var getValidInputValue = useCallback(function (inputValue) {
        var number = parseInt(inputValue, 10);
        return inputValue === '' || Number.isNaN(number) ? '' : Math.abs(number);
    }, []);
    var handleSliderChange = useCallback(function (event, payload) {
        if (onChange)
            onChange(event, payload);
        if (onSliderChange)
            onSliderChange(event, payload);
    }, [onChange, onSliderChange]);
    var handleInputChange = useCallback(function (event, payload) {
        if (onChange)
            onChange(event, { value: getValidInputValue(payload.value) });
        if (onInputChange)
            onInputChange(event, { value: getValidInputValue(payload.value) });
    }, [getValidInputValue, onChange, onInputChange]);
    return (React.createElement("div", { className: cn(styles.component, (_b = {},
            _b[styles.block] = block,
            _b[styles.filled] = Boolean(value),
            _b[styles.hasLabel] = label,
            _b[styles.hasError] = Boolean(error),
            _b), styles[size], className), "data-test-id": dataTestId },
        React.createElement(Input$1, __assign({}, restProps, customInputProps, { ref: ref, value: value.toString(), onChange: handleInputChange, block: true, size: size, label: label, disabled: disabled, readOnly: readOnly, className: cn(inputClassName, styles.input), focusedClassName: cn(focusedClassName, styles.focused), fieldClassName: cn(fieldClassName, styles.field), inputMode: 'numeric', pattern: '[0-9]*', error: error, bottomAddons: !disabled && (React.createElement(Slider, { min: min, max: max, step: step, onChange: handleSliderChange, ref: ref, value: Number.isNaN(sliderValue) ? 0 : sliderValue, disabled: disabled || readOnly, className: cn(styles.slider, sliderClassName) })), rightAddons: (info || rightAddons) && (React.createElement(Fragment, null,
                info && React.createElement("span", { className: styles.info }, info),
                rightAddons)) })),
        steps.length > 0 && !error && (React.createElement("div", { className: cn(styles.steps, stepsClassName) }, steps.map(function (stepLabel, i) {
            return isValidElement(stepLabel) ? (cloneElement(stepLabel, { key: i })) : (
            // eslint-disable-next-line react/no-array-index-key
            React.createElement("span", { key: i }, stepLabel));
        })))));
});
/**
 * Для отображения в сторибуке
 */
SliderInput.defaultProps = {
    value: '',
    min: 0,
    max: 100,
    step: 1,
    steps: [],
    size: 's',
    Input: Input,
    customInputProps: {},
};

export { SliderInput };
