// @alfalab/core-components/gallery/cssm/Component.js
Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var coreComponentsBaseModal = require('../../base-modal/cssm');
require('classnames');
require('@alfalab/hooks');
var context = require('./context.js');
require('./components/image-preview/index.module.css');
require('./components/image-preview/Component.js');
require('./components/navigation-bar/index.module.css');
require('./utils/split-filename.js');
require('./utils/utils.js');
require('./utils/constants.js');
var components_navigationBar_Component = require('./components/navigation-bar/Component.js');
require('../../typography/cssm');
require('./components/header-info-block/index.module.css');
require('./components/header-info-block/Component.js');
require('./tslib.es6-ce870b46.js');
require('../../icon-button/cssm');
require('../../tooltip/cssm');
require('@alfalab/icons-glyph/PointerDownMIcon');
require('@alfalab/icons-glyph/ArrowsOutwardMIcon');
require('@alfalab/icons-glyph/ArrowsInwardMIcon');
require('@alfalab/icons-glyph/CrossMIcon');
require('./components/header/buttons.js');
require('./components/header/index.module.css');
var components_header_Component = require('./components/header/Component.js');
require('swiper');
require('swiper/react');
require('element-closest');
require('swiper/swiper.min.css');
require('@alfalab/icons-glyph/ChevronBackHeavyMIcon');
require('@alfalab/icons-glyph/ChevronForwardHeavyMIcon');
require('./components/image-viewer/index.module.css');
require('./components/image-viewer/slide.js');
var components_imageViewer_component = require('./components/image-viewer/component.js');
var styles = require('./index.module.css');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);
var styles__default = /*#__PURE__*/_interopDefaultLegacy(styles);

var Backdrop = function () { return null; };
var Gallery = function (_a) {
    var open = _a.open, images = _a.images, _b = _a.initialSlide, initialSlide = _b === void 0 ? 0 : _b, _c = _a.loop, loop = _c === void 0 ? true : _c, onClose = _a.onClose;
    var _d = React.useState(initialSlide), currentSlideIndex = _d[0], setCurrentSlideIndex = _d[1];
    var _e = React.useState(), swiper = _e[0], setSwiper = _e[1];
    var _f = React.useState([]), imagesMeta = _f[0], setImagesMeta = _f[1];
    var _g = React.useState(false), fullScreen = _g[0], setFullScreen = _g[1];
    var slideTo = React.useCallback(function (index) {
        if (images[index]) {
            setCurrentSlideIndex(index);
            if (swiper) {
                swiper.slideTo(index);
            }
        }
    }, [images, swiper]);
    var slideNext = React.useCallback(function () {
        var lastIndex = images.length - 1;
        var nextIndex = currentSlideIndex + 1;
        if (nextIndex >= images.length) {
            nextIndex = loop ? 0 : lastIndex;
        }
        slideTo(nextIndex);
    }, [images.length, loop, currentSlideIndex, slideTo]);
    var slidePrev = React.useCallback(function () {
        var lastIndex = images.length - 1;
        var nextIndex = currentSlideIndex - 1;
        if (nextIndex < 0) {
            nextIndex = loop ? lastIndex : 0;
        }
        slideTo(nextIndex);
    }, [images.length, loop, currentSlideIndex, slideTo]);
    var setImageMeta = React.useCallback(function (meta, index) {
        imagesMeta[index] = meta;
        setImagesMeta(imagesMeta.slice());
    }, [imagesMeta]);
    var handleClose = React.useCallback(function () {
        onClose();
        setCurrentSlideIndex(initialSlide);
        setFullScreen(false);
    }, [initialSlide, onClose]);
    var handleEscapeKeyDown = function () {
        if (fullScreen) {
            setFullScreen(false);
        }
        else {
            handleClose();
        }
    };
    var handleKeyDown = React.useCallback(function (event) {
        if (!open || fullScreen) {
            return;
        }
        switch (event.key) {
            case 'ArrowLeft':
                slidePrev();
                break;
            case 'ArrowRight':
                slideNext();
                break;
        }
    }, [fullScreen, open, slideNext, slidePrev]);
    React.useEffect(function () {
        document.addEventListener('keydown', handleKeyDown);
        return function () {
            document.removeEventListener('keydown', handleKeyDown);
        };
    }, [handleKeyDown]);
    var singleSlide = images.length === 1;
    var showNavigationBar = !singleSlide && !fullScreen;
    var galleryContext = {
        singleSlide: singleSlide,
        currentSlideIndex: currentSlideIndex,
        images: images,
        imagesMeta: imagesMeta,
        fullScreen: fullScreen,
        initialSlide: initialSlide,
        setFullScreen: setFullScreen,
        setImageMeta: setImageMeta,
        slideNext: slideNext,
        slidePrev: slidePrev,
        slideTo: slideTo,
        getSwiper: function () { return swiper; },
        setSwiper: setSwiper,
        onClose: handleClose,
        setCurrentSlideIndex: setCurrentSlideIndex,
        getCurrentImage: function () { return images[currentSlideIndex]; },
        getCurrentImageMeta: function () { return imagesMeta[currentSlideIndex]; },
    };
    return (React__default['default'].createElement(context.GalleryContext.Provider, { value: galleryContext },
        React__default['default'].createElement(coreComponentsBaseModal.BaseModal, { open: open, className: styles__default['default'].modal, onEscapeKeyDown: handleEscapeKeyDown, Backdrop: Backdrop },
            React__default['default'].createElement("div", { className: styles__default['default'].container },
                React__default['default'].createElement(components_header_Component.Header, null),
                React__default['default'].createElement(components_imageViewer_component.ImageViewer, null),
                showNavigationBar && React__default['default'].createElement(components_navigationBar_Component.NavigationBar, null)))));
};

exports.Gallery = Gallery;
