// @alfalab/core-components/tabs/modern/responsive.js
import 'react';
import 'classnames';
import 'compute-scroll-into-view';
import '@alfalab/hooks';
import '../../keyboard-focusable/modern';
export { ScrollableContainer } from './components/scrollable-container/Component.js';
export { useTabs } from './useTabs.js';
import './components/primary-tablist/Component.js';
import './index.module-c9e59889.js';
import './components/primary-tablist/Component.desktop.js';
import '../../tag/modern';
import './components/secondary-tablist/Component.js';
import './index.module-f5e5a0a0.js';
import './components/secondary-tablist/Component.desktop.js';
export { Tab } from './components/tab/Component.js';
import './components/tabs/Component.js';
import './components/primary-tablist/Component.mobile.js';
export { PrimaryTabListResponsive } from './components/primary-tablist/Component.responsive.js';
import './components/secondary-tablist/Component.mobile.js';
export { SecondaryTabListResponsive } from './components/secondary-tablist/Component.responsive.js';
export { TabsResponsive } from './components/tabs/Component.responsive.js';
