// @alfalab/core-components/chart/modern/hooks/useSettings/utils/setGradientCharts.js
const setGradientCharts = (series) => {
    const filterSeries = series.filter(item => item.chart !== 'gradient');
    return filterSeries.reduce((accum, item) => {
        const { chart, data: dataSeria, offset, fill } = item;
        if (chart === 'area') {
            let newData = null;
            if (offset) {
                newData = dataSeria.map((d) => {
                    const { label, value } = d;
                    return {
                        label,
                        value: Math.ceil(value - value * offset),
                    };
                });
            }
            accum.push({
                ...item,
                zIndex: -100,
                chart: 'gradient',
                hideLegend: true,
                hideTooltip: true,
                properties: {
                    ...item.properties,
                    dataKey: `${item.properties.dataKey}-gradient`,
                    fill,
                },
                data: newData || item.data,
            });
        }
        accum.push(item);
        return accum;
    }, []);
};

export { setGradientCharts };
