// @alfalab/core-components/gallery/esm/components/image-viewer/component.js
import React, { useContext, useRef, useCallback, useEffect, useMemo } from 'react';
import cn from 'classnames';
import { useFocus } from '@alfalab/hooks';
import { GalleryContext } from '../../context.js';
import { getImageAlt, getImageKey } from '../../utils/utils.js';
import { TestIds } from '../../utils/constants.js';
import '../../../../typography/esm';
import { _ as __assign } from '../../tslib.es6-6bc7eb3c.js';
import SwiperCore, { EffectFade, A11y, Controller } from 'swiper';
import { Swiper, SwiperSlide } from 'swiper/react';
import elementClosest from 'element-closest';
import 'swiper/swiper.min.css';
import { ChevronBackHeavyMIcon } from '@alfalab/icons-glyph/ChevronBackHeavyMIcon';
import { ChevronForwardHeavyMIcon } from '@alfalab/icons-glyph/ChevronForwardHeavyMIcon';
import { s as styles, S as Slide } from '../../slide-438a9380.js';

SwiperCore.use([EffectFade, A11y, Controller]);
var ImageViewer = function () {
    var _a, _b, _c;
    var _d = useContext(GalleryContext), singleSlide = _d.singleSlide, images = _d.images, imagesMeta = _d.imagesMeta, fullScreen = _d.fullScreen, currentSlideIndex = _d.currentSlideIndex, initialSlide = _d.initialSlide, onClose = _d.onClose, getCurrentImage = _d.getCurrentImage, setImageMeta = _d.setImageMeta, setCurrentSlideIndex = _d.setCurrentSlideIndex, getSwiper = _d.getSwiper, setSwiper = _d.setSwiper, slidePrev = _d.slidePrev, slideNext = _d.slideNext;
    var leftArrowRef = useRef(null);
    var rightArrowRef = useRef(null);
    var leftArrowFocused = useFocus(leftArrowRef, 'keyboard')[0];
    var rightArrowFocused = useFocus(rightArrowRef, 'keyboard')[0];
    var swiper = getSwiper();
    var handleSlideChange = useCallback(function () {
        var _a;
        setCurrentSlideIndex((_a = swiper === null || swiper === void 0 ? void 0 : swiper.activeIndex) !== null && _a !== void 0 ? _a : initialSlide);
    }, [setCurrentSlideIndex, swiper, initialSlide]);
    var handlePrevClick = function () {
        slidePrev();
    };
    var handleNextClick = function () {
        slideNext();
    };
    var handleArrowLeftKeyDown = function (event) {
        if (event.key === 'Enter') {
            slidePrev();
        }
    };
    var handleArrowRightKeyDown = function (event) {
        if (event.key === 'Enter') {
            slideNext();
        }
    };
    var handleLoad = function (event, index) {
        var target = event.currentTarget;
        var naturalWidth = target.naturalWidth, naturalHeight = target.naturalHeight;
        setImageMeta({ width: naturalWidth, height: naturalHeight }, index);
    };
    var handleLoadError = function (index) {
        setImageMeta({ width: 0, height: 0, broken: true }, index);
    };
    var handleWrapperClick = useCallback(function (event) {
        var _a, _b;
        var eventTarget = event.target;
        var isArrow = ((_a = leftArrowRef.current) === null || _a === void 0 ? void 0 : _a.contains(eventTarget)) || ((_b = rightArrowRef.current) === null || _b === void 0 ? void 0 : _b.contains(eventTarget));
        var isPlaceholder = Boolean(eventTarget.closest("." + styles.placeholder));
        var isImg = eventTarget.tagName === 'IMG';
        if (!isImg && !isPlaceholder && !isArrow) {
            onClose();
        }
    }, [onClose]);
    useEffect(function () {
        elementClosest(window);
    }, []);
    var swiperProps = useMemo(function () {
        var _a;
        return ({
            slidesPerView: 1,
            effect: 'fade',
            fadeEffect: {
                crossFade: true,
            },
            className: cn(styles.swiper, (_a = {}, _a[styles.hidden] = fullScreen, _a)),
            controller: { control: swiper },
            a11y: {
                slideRole: 'img',
            },
            initialSlide: initialSlide,
            simulateTouch: false,
            onSwiper: setSwiper,
            onSlideChange: handleSlideChange,
        });
    }, [swiper, fullScreen, initialSlide, handleSlideChange, setSwiper]);
    var showControls = !singleSlide && !fullScreen;
    var swiperWidth = (swiper === null || swiper === void 0 ? void 0 : swiper.width) || 1;
    var swiperHeight = (swiper === null || swiper === void 0 ? void 0 : swiper.height) || 1;
    var swiperAspectRatio = swiperWidth / swiperHeight;
    var currentImage = getCurrentImage();
    return (
    /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-static-element-interactions */
    React.createElement("div", { className: cn(styles.component, (_a = {}, _a[styles.singleSlide] = singleSlide, _a)), onClick: handleWrapperClick },
        showControls && (React.createElement("div", { className: cn(styles.arrow, (_b = {},
                _b[styles.focused] = leftArrowFocused,
                _b)), onClick: handlePrevClick, role: 'button', onKeyDown: handleArrowLeftKeyDown, tabIndex: 0, ref: leftArrowRef, "aria-label": '\u041F\u0440\u0435\u0434\u044B\u0434\u0443\u0449\u0435\u0435 \u0438\u0437\u043E\u0431\u0440\u0430\u0436\u0435\u043D\u0438\u0435', "data-test-id": TestIds.PREV_SLIDE_BUTTON },
            React.createElement(ChevronBackHeavyMIcon, null))),
        fullScreen && (React.createElement("img", { src: currentImage === null || currentImage === void 0 ? void 0 : currentImage.src, alt: currentImage ? getImageAlt(currentImage, currentSlideIndex) : '', className: styles.fullScreenImage })),
        React.createElement(Swiper, __assign({}, swiperProps), images.map(function (image, index) {
            var meta = imagesMeta[index];
            var imageWidth = (meta === null || meta === void 0 ? void 0 : meta.width) || 1;
            var imageHeight = (meta === null || meta === void 0 ? void 0 : meta.height) || 1;
            var imageAspectRatio = imageWidth / imageHeight;
            var slideVisible = index === currentSlideIndex;
            return (React.createElement(SwiperSlide, { key: getImageKey(image, index), style: {
                    pointerEvents: slideVisible ? 'auto' : 'none',
                    transitionProperty: 'opacity',
                } }, function (_a) {
                var isActive = _a.isActive;
                return (React.createElement(Slide, { isActive: isActive, swiperAspectRatio: swiperAspectRatio, image: image, swiperHeight: swiperHeight, meta: meta, index: index, imageAspectRatio: imageAspectRatio, slideVisible: slideVisible, handleLoad: handleLoad, handleLoadError: handleLoadError }));
            }));
        })),
        showControls && (React.createElement("div", { className: cn(styles.arrow, (_c = {},
                _c[styles.focused] = rightArrowFocused,
                _c)), onClick: handleNextClick, role: 'button', onKeyDown: handleArrowRightKeyDown, tabIndex: 0, ref: rightArrowRef, "aria-label": '\u0421\u043B\u0435\u0434\u0443\u044E\u0449\u0435\u0435 \u0438\u0437\u043E\u0431\u0440\u0430\u0436\u0435\u043D\u0438\u0435', "data-test-id": TestIds.NEXT_SLIDE_BUTTON },
            React.createElement(ChevronForwardHeavyMIcon, null)))));
};

export { ImageViewer };
