// @alfalab/core-components/tooltip/esm/Component-72ed3da9.js
import React, { useState, useRef, useCallback, useEffect, Fragment } from 'react';
import cn from 'classnames';
import mergeRefs from 'react-merge-refs';
import { Popover } from '../../popover/esm';

/*! *****************************************************************************
Copyright (c) Microsoft Corporation.

Permission to use, copy, modify, and/or distribute this software for any
purpose with or without fee is hereby granted.

THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES WITH
REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY
AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY SPECIAL, DIRECT,
INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM
LOSS OF USE, DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION WITH THE USE OR
PERFORMANCE OF THIS SOFTWARE.
***************************************************************************** */
var __assign = function () {
    __assign = Object.assign || function __assign(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s)
                if (Object.prototype.hasOwnProperty.call(s, p))
                    t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
function __rest(s, e) {
    var t = {};
    for (var p in s)
        if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
            t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) {
            if (e.indexOf(p[i]) < 0 && Object.prototype.propertyIsEnumerable.call(s, p[i]))
                t[p[i]] = s[p[i]];
        }
    return t;
}

var styles = {"component":"tooltip__component_19vu5","popper":"tooltip__popper_19vu5","hint":"tooltip__hint_19vu5","arrow":"tooltip__arrow_19vu5","target":"tooltip__target_19vu5","overlap":"tooltip__overlap_19vu5"};
require('./index.css');

var Tooltip = function (_a) {
    var children = _a.children, content = _a.content, _b = _a.trigger, trigger = _b === void 0 ? 'hover' : _b, _c = _a.onCloseDelay, onCloseDelay = _c === void 0 ? 300 : _c, _d = _a.onOpenDelay, onOpenDelay = _d === void 0 ? 300 : _d, dataTestId = _a.dataTestId, forcedOpen = _a.open, _e = _a.offset, offset = _e === void 0 ? [0, 16] : _e, position = _a.position, contentClassName = _a.contentClassName, arrowClassName = _a.arrowClassName, popoverClassName = _a.popoverClassName, updatePopover = _a.updatePopover, targetClassName = _a.targetClassName, zIndex = _a.zIndex, onClose = _a.onClose, onOpen = _a.onOpen, getPortalContainer = _a.getPortalContainer, _f = _a.view, view = _f === void 0 ? 'tooltip' : _f, _g = _a.targetRef, targetRef = _g === void 0 ? null : _g, fallbackPlacements = _a.fallbackPlacements, _h = _a.preventOverflow, preventOverflow = _h === void 0 ? true : _h, _j = _a.availableHeight, availableHeight = _j === void 0 ? false : _j;
    var _k = useState(!!forcedOpen), visible = _k[0], setVisible = _k[1];
    var _l = useState(null), target = _l[0], setTarget = _l[1];
    var contentRef = useRef(null);
    var timer = useRef(0);
    var show = forcedOpen === undefined ? visible : forcedOpen;
    var open = function () {
        if (!show) {
            setVisible(true);
            if (onOpen) {
                onOpen();
            }
        }
    };
    var close = useCallback(function () {
        if (show) {
            setVisible(false);
            if (onClose) {
                onClose();
            }
        }
    }, [onClose, show]);
    var toggle = function () {
        if (show) {
            close();
        }
        else {
            open();
        }
    };
    var clickedOutside = useCallback(function (node) {
        if (target && target.contains(node)) {
            return false;
        }
        if (contentRef.current && contentRef.current.contains(node)) {
            return false;
        }
        return true;
    }, [target]);
    useEffect(function () {
        var handleBodyClick = function (event) {
            var eventTarget = event.target;
            if (clickedOutside(eventTarget)) {
                close();
            }
        };
        document.body.addEventListener('click', handleBodyClick);
        return function () {
            document.body.removeEventListener('click', handleBodyClick);
            clearTimeout(timer.current);
        };
    }, [clickedOutside, close]);
    var handleTargetClick = function () {
        toggle();
    };
    var handleMouseOver = function () {
        clearTimeout(timer.current);
        timer.current = window.setTimeout(function () {
            open();
        }, onOpenDelay);
    };
    var handleMouseOut = function () {
        clearTimeout(timer.current);
        timer.current = window.setTimeout(function () {
            close();
        }, onCloseDelay);
    };
    var handleTouchStart = function (event) {
        var eventTarget = event.target;
        clearTimeout(timer.current);
        if (clickedOutside(eventTarget)) {
            timer.current = window.setTimeout(function () {
                close();
            }, onCloseDelay);
        }
        else {
            open();
        }
    };
    var onClickProps = { onClick: handleTargetClick };
    var onHoverProps = {
        onMouseOver: handleMouseOver,
        onMouseOut: handleMouseOut,
        onTouchStart: handleTouchStart,
    };
    var getTargetProps = function () {
        var props = {
            className: cn(styles.target, targetClassName),
        };
        switch (trigger) {
            case 'click':
                return __assign(__assign({}, props), onClickProps);
            case 'hover':
                return __assign(__assign({}, props), onHoverProps);
            default:
                return {};
        }
    };
    var getContentProps = function () {
        var props = {
            ref: contentRef,
            'data-test-id': dataTestId,
            className: cn(styles.component, contentClassName),
        };
        switch (trigger) {
            case 'hover':
                return __assign(__assign({}, props), onHoverProps);
            default:
                return props;
        }
    };
    return (React.createElement(Fragment, null,
        React.createElement("div", __assign({ ref: mergeRefs([targetRef, setTarget]) }, getTargetProps()),
            children.props.disabled && React.createElement("div", { className: styles.overlap }),
            children),
        React.createElement(Popover, { anchorElement: target, open: show, getPortalContainer: getPortalContainer, arrowClassName: cn(arrowClassName, styles.arrow), popperClassName: cn(styles.popper, styles[view]), className: popoverClassName, offset: offset, withArrow: true, position: position, update: updatePopover, zIndex: zIndex, fallbackPlacements: fallbackPlacements, preventOverflow: preventOverflow, availableHeight: availableHeight },
            React.createElement("div", __assign({}, getContentProps()), content))));
};

export { Tooltip as T, __rest as _, __assign as a };
